### PR TITLE
nm.plugin: Checking libnm/NM version mismatch

### PR DESCRIPTION
**Pull request by UnversedBlood**
Added in libnmstate/nm/plugin.py changes that decodes and check the
version mismatch when the plugin class is instantiated.

Signed-off-by: Víctor Javier Díaz Garrido <vicdigar@hotmail.com>

### CODE REVIEW

File: libnmstate/nm/plugin.py [hunk 1]
```diff
@@ -43,6 +46,7 @@
 
 class NetworkManagerPlugin:
     def __init__(self):
+        self.check_version_mismatch()
         self._ctx = NmContext()
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:50]: This is wrong. It should be:
```
self._ctx = NmContext()
self._check_version_mismatch()
```
This way the context is initialized and you are going to be able to do `self._ctx.client()` in `check_version_mismatch()`. Please, see the comments below.

File: libnmstate/nm/plugin.py [hunk 2]
```diff
@@ -17,6 +17,9 @@
 # along with this program. If not, see <https://www.gnu.org/licenses/>.
 #
 from operator import itemgetter
+from warnings import warn
+from distutils.version import StrictVersion
+from gi.repository import NM
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:22]: Please, sort the imports by alphabetical order.

File: libnmstate/nm/plugin.py [hunk 3]
```diff
@@ -140,6 +144,18 @@ def get_dns_client_config(self):
     def refresh_content(self):
         self._ctx.refresh_content()
 
+    @staticmethod
+    def check_version_mismatch():
+        nm_client_version = _nm_client_decode_version()
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:149]: This should be `nm_client_version = _nm_client_decode_version(self._ctx.client())`.

File: libnmstate/nm/plugin.py [hunk 4]
```diff
@@ -159,3 +175,21 @@ def _remove_ovs_bridge_unsupported_entries(iface_info):
...
+        + "."
+        + str((v >> 8) & 0xFF)
+        + "."
+        + str((v) & 0xFF)
+    )
+
+
+def _nm_client_decode_version(v=None):
+    if v is None:
+        v = NM.Client.get_version()
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:194]: You cannot do this. You need a client instance that you should get from `_check_version_mismatch()`. Then, you can do: `v = nm_client.get_version()`.

File: libnmstate/nm/plugin.py [hunk 5]
```diff
@@ -159,3 +175,21 @@ def _remove_ovs_bridge_unsupported_entries(iface_info):
...
+    return (
+        str((v >> 16) & 0xFF)
+        + "."
+        + str((v >> 8) & 0xFF)
+        + "."
+        + str((v) & 0xFF)
+    )
+
+
+def _nm_client_decode_version(v=None):
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:192]: `v=None` is not needed here.

File: libnmstate/nm/plugin.py [hunk 6]
```diff
@@ -159,3 +175,21 @@ def _remove_ovs_bridge_unsupported_entries(iface_info):
     iface_info.pop(Interface.MTU, None)
 
     return iface_info
+
+
+def _nm_utils_decode_version(v=None):
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:180]: `v=None` is not needed here.

File: libnmstate/nm/plugin.py [hunk 7]
```diff
@@ -140,6 +144,18 @@ def get_dns_client_config(self):
...
+    def check_version_mismatch():
+        nm_client_version = _nm_client_decode_version()
+        nm_utils_version = _nm_utils_decode_version()
+
+        if nm_client_version is None:
+            warn("NetworkManager is not running\n")
+        elif StrictVersion(nm_client_version) != StrictVersion(
+            nm_utils_version
+        ):
+            warn("Version mismatch\n")
```
Comment (ffmancera) [inline libnmstate/nm/plugin.py:157]: It would be good to have a log like "libnm version XXXX mismatches NetworkManager version YYYY"

**Comment by UnversedBlood** [comment]
> 
> 
> Thanks for your contribution, please see the comments below. Let me know if you have any question!

Thanks to you for the comments, I have fixed what you have indicated.

File: libnmstate/nm/plugin.py (hunk 3 above)
Comment (ffmancera) [reply libnmstate/nm/plugin.py]: My bad. It should be `self._ctx.client`.

Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done!

File: libnmstate/nm/plugin.py (hunk 1 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thank you!

File: libnmstate/nm/plugin.py (hunk 2 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thank you!

File: libnmstate/nm/plugin.py (hunk 5 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thank you!

File: libnmstate/nm/plugin.py (hunk 4 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thank you!

File: libnmstate/nm/plugin.py (hunk 6 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thank you!

File: libnmstate/nm/plugin.py (hunk 7 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thank you!

File: libnmstate/nm/plugin.py [hunk 8]
```diff
@@ -16,7 +16,12 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with this program. If not, see <https://www.gnu.org/licenses/>.
 #
+from distutils.version import StrictVersion
 from operator import itemgetter
+from warnings import warn
```
Comment (cathay4t) [inline libnmstate/nm/plugin.py:21]: I would prefer `logging.warning` where API user could control the output target of these warnings.
Unlike CLI tools, the library should not print to stderr unless asked explicitly.

Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done!

File: libnmstate/nm/plugin.py [hunk 9]
```diff
@@ -159,3 +180,19 @@ def _remove_ovs_bridge_unsupported_entries(iface_info):
...
+
+def _nm_utils_decode_version():
+    v = NM.utils_version()
+    return (
+        str((v >> 16) & 0xFF)
+        + "."
+        + str((v >> 8) & 0xFF)
+        + "."
+        + str((v) & 0xFF)
+    )
```
Comment (cathay4t) [inline libnmstate/nm/plugin.py:193]: Please use `NM.MAJOR_VERSION`, `NM.MINOR_VERSION` and `NM.MICRO_VERSION`.

File: libnmstate/nm/plugin.py [hunk 10]
```diff
@@ -140,6 +146,21 @@ def get_dns_client_config(self):
     def refresh_content(self):
         self._ctx.refresh_content()
 
+    def _check_version_mismatch(self):
+        nm_client_version = _nm_client_decode_version(self._ctx.client)
```
Comment (cathay4t) [inline libnmstate/nm/plugin.py:150]: `nm_client_version = self._ctx.client.get_version()` is enough. No need to wrap one line code as function.

File: libnmstate/nm/plugin.py (hunk 9 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py:193]: Done, thank you!

File: libnmstate/nm/plugin.py (hunk 10 above)
Comment (UnversedBlood) [reply libnmstate/nm/plugin.py:150]: You are right, thanks!

**Comment by ffmancera** [comment]
This requires your ack @cathay4t, thanks!

File: libnmstate/nm/plugin.py [hunk 11]
```diff
@@ -161,3 +182,13 @@ def _remove_ovs_bridge_unsupported_entries(iface_info):
...
+
+
+def _nm_utils_decode_version():
+    return (
+        str(NM.MAJOR_VERSION)
+        + "."
+        + str(NM.MICRO_VERSION)
+        + "."
+        + str(NM.MINOR_VERSION)
+    )
```
Comment (cathay4t) [inline libnmstate/nm/plugin.py:194]: This looks ugly.
Try:

```python
f"{NM.MAJOR_VERSION}.{NM.MINOR_VERSION}.{NM.MICRO_VERSION}"
```

Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done, thanks for the tip!

File: libnmstate/nm/plugin.py [hunk 12]
```diff
@@ -16,8 +16,13 @@
 # You should have received a copy of the GNU Lesser General Public License
 # along with this program. If not, see <https://www.gnu.org/licenses/>.
 #
+from distutils.version import StrictVersion
+import logging
 from operator import itemgetter
 
+from gi.repository import NM
```
Comment (cathay4t) [inline libnmstate/nm/plugin.py:23]: Please use `from .common import NM`

Comment (UnversedBlood) [reply libnmstate/nm/plugin.py]: Done!

**Comment by cathay4t** [comment]
@UnversedBlood Please resolve the conflict. Then we can merge.

**Comment by UnversedBlood** [comment]
> @UnversedBlood Please resolve the conflict. Then we can merge.

Conflict resolved!

**Comment by cathay4t** [comment]
@UnversedBlood Thanks for the contribution.

**Pull request merged**